runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (C) 2009-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.grpc.scaladsl
+
+import akka.NotUsed
+import akka.annotation.{ DoNotInherit, InternalApi }
+import akka.grpc.internal.ChannelApiHelpers
+import akka.stream.{ Materializer, OverflowStrategy }
+import akka.stream.scaladsl.{ Sink, Source }
+import io.grpc.stub.{ ClientCalls, StreamObserver }
+import io.grpc.{ CallOptions, Channel, Deadline, MethodDescriptor }
+
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * Request builder for requests giving providing per call metadata capabilities
+ *
+ * Not for user extension
+ */
+@DoNotInherit
+sealed trait RequestBuilder[Req, Res] {
+
+  def addMetadata(key: String, value: String): RequestBuilder[Req, Res]
+  def withDeadline(deadline: FiniteDuration): RequestBuilder[Req, Res]
+  def execute(req: Req): Res
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+final case class RequestBuilderImpl[Req, Res, Ret](
+  options: CallOptions,
+  delegatedExecute: (Req, CallOptions) => Ret) extends RequestBuilder[Req, Ret]() {
+
+  def addMetadata(key: String, value: String): RequestBuilderImpl[Req, Res, Ret] = {
+    // FIXME support values other than string values?
+    copy(options = options.withOption(CallOptions.Key.of(key, ""), value))
+  }
+
+  def withDeadline(deadline: FiniteDuration): RequestBuilderImpl[Req, Res, Ret] = {
+    copy(options = options.withDeadline(Deadline.after(deadline.length, deadline.unit)))
```

Comment:
Nice, we don't support this at the server side yet (#188) but since this is :+1:

Reply:
Yeah, just picked this randomly to have some functionality to provide through the request builder.